File: solana-escrow/programs/polybaskets-escrow/src/math.rs
use anchor_lang::prelude::*;

use crate::constants::{MANAGEMENT_FEE_BPS, MANAGEMENT_FEE_PERIOD_SECS, MAX_BPS, SHARE_SCALE};
use crate::errors::EscrowError;

/// Protocol fees round up so dust-sized requests cannot avoid configured fees.
pub fn fee_ceil(amount: u64, fee_bps: u16) -> Result<u64> {
    let numerator = (amount as u128)
        .checked_mul(fee_bps as u128)
        .and_then(|value| value.checked_add(MAX_BPS as u128 - 1))
        .ok_or(EscrowError::MathOverflow)?;
    let fee = numerator
        .checked_div(MAX_BPS as u128)
        .ok_or(EscrowError::MathOverflow)?;
    u64::try_from(fee).map_err(|_| error!(EscrowError::MathOverflow))
}

/// Deposit shares round down in the protocol's favour.
pub fn shares_for_value(net_value: u64, share_price: u64) -> Result<u64> {
    require!(share_price > 0, EscrowError::InvalidSettlementValues);
    let shares = (net_value as u128)
        .checked_mul(SHARE_SCALE as u128)
        .and_then(|value| value.checked_div(share_price as u128))
        .ok_or(EscrowError::MathOverflow)?;
    u64::try_from(shares).map_err(|_| error!(EscrowError::MathOverflow))
}

/// Current share price from the backend-computed NAV and on-chain supply.
pub fn share_price_from_nav(nav_value: u64, total_shares: u64) -> Result<u64> {
    require!(total_shares > 0, EscrowError::InvalidSettlementValues);
    let price = (nav_value as u128)
        .checked_mul(SHARE_SCALE as u128)
        .and_then(|value| value.checked_div(total_shares as u128))
        .ok_or(EscrowError::MathOverflow)?;
    u64::try_from(price).map_err(|_| error!(EscrowError::MathOverflow))
}

/// Gross value represented by shares at a six-decimal share price.
pub fn value_for_shares(shares: u64, share_price: u64) -> Result<u64> {
    let value = (shares as u128)
        .checked_mul(share_price as u128)
        .and_then(|amount| amount.checked_div(SHARE_SCALE as u128))
        .ok_or(EscrowError::MathOverflow)?;
    u64::try_from(value).map_err(|_| error!(EscrowError::MathOverflow))
}

/// Shares minted to transfer exactly 0.35% of user AUM for one fee period.
pub fn management_fee_shares(total_shares: u64) -> Result<u64> {
    let denominator = MAX_BPS
        .checked_sub(MANAGEMENT_FEE_BPS)
        .ok_or(EscrowError::InvalidBasisPoints)?;
    let shares = (total_shares as u128)
        .checked_mul(MANAGEMENT_FEE_BPS as u128)
        .and_then(|value| value.checked_div(denominator as u128))
        .ok_or(EscrowError::MathOverflow)?;
    u64::try_from(shares).map_err(|_| error!(EscrowError::MathOverflow))
}

/// Time-weights monthly management dilution and carries sub-share precision
/// forward so frequent keeper calls cannot erase fees through rounding.
pub fn management_fee_shares_for_elapsed(
    total_shares: u64,
    elapsed_seconds: i64,
    prior_remainder: u128,
) -> Result<(u64, u128)> {
    require!(elapsed_seconds >= 0, EscrowError::InvalidSettlementValues);
    let elapsed = u128::try_from(elapsed_seconds).map_err(|_| EscrowError::MathOverflow)?;
    let period =
        u128::try_from(MANAGEMENT_FEE_PERIOD_SECS).map_err(|_| EscrowError::MathOverflow)?;
    let dilution_denominator = (MAX_BPS as u128)
        .checked_sub(MANAGEMENT_FEE_BPS as u128)
        .and_then(|value| value.checked_mul(period))
        .ok_or(EscrowError::MathOverflow)?;
    require!(
        prior_remainder < dilution_denominator,
        EscrowError::InvalidSettlementValues
    );
    let numerator = (total_shares as u128)
        .checked_mul(MANAGEMENT_FEE_BPS as u128)
        .and_then(|value| value.checked_mul(elapsed))
        .and_then(|value| value.checked_add(prior_remainder))
        .ok_or(EscrowError::MathOverflow)?;
    let minted = numerator
        .checked_div(dilution_denominator)
        .ok_or(EscrowError::MathOverflow)?;
    let remainder = numerator
        .checked_rem(dilution_denominator)
        .ok_or(EscrowError::MathOverflow)?;
    Ok((
        u64::try_from(minted).map_err(|_| error!(EscrowError::MathOverflow))?,
        remainder,
    ))
}

/// Performance fees round down so a creator can never receive more than the
/// configured percentage of realized profit.
pub fn fee_floor(amount: u64, fee_bps: u16) -> Result<u64> {
    let fee = (amount as u128)
        .checked_mul(fee_bps as u128)
        .and_then(|value| value.checked_div(MAX_BPS as u128))
        .ok_or(EscrowError::MathOverflow)?;
    u64::try_from(fee).map_err(|_| error!(EscrowError::MathOverflow))
}

/// Allocates cost basis to a partial redemption. Partial allocations round up
/// in the user's favour; a full redemption consumes the exact remaining basis.
pub fn cost_basis_for_shares(
    cost_basis: u64,
    shares_redeemed: u64,
    shares_owned: u64,
) -> Result<u64> {
    require!(
        shares_owned > 0 && shares_redeemed <= shares_owned,
        EscrowError::InvalidSettlementValues
    );
    if shares_redeemed == shares_owned {
        return Ok(cost_basis);
    }
    let numerator = (cost_basis as u128)
        .checked_mul(shares_redeemed as u128)
        .and_then(|value| value.checked_add(shares_owned as u128 - 1))
        .ok_or(EscrowError::MathOverflow)?;
    let basis = numerator
        .checked_div(shares_owned as u128)
        .ok_or(EscrowError::MathOverflow)?;
    u64::try_from(basis).map_err(|_| error!(EscrowError::MathOverflow))
}

/// Minimum acceptable output derived from a quoted output and tolerance.
pub fn minimum_after_slippage(quoted_out: u64, tolerance_bps: u16) -> Result<u64> {
    let retained_bps = MAX_BPS
        .checked_sub(tolerance_bps)
        .ok_or(EscrowError::InvalidBasisPoints)?;
    let minimum = (quoted_out as u128)
        .checked_mul(retained_bps as u128)
        .and_then(|value| value.checked_div(MAX_BPS as u128))
        .ok_or(EscrowError::MathOverflow)?;
    u64::try_from(minimum).map_err(|_| error!(EscrowError::MathOverflow))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fee_rounds_up() {
        assert_eq!(fee_ceil(100_000_000, 50).unwrap(), 500_000);
        assert_eq!(fee_ceil(1, 50).unwrap(), 1);
    }

    #[test]
    fn shares_round_down() {
        assert_eq!(
            shares_for_value(494_000_000, 1_000_000).unwrap(),
            494_000_000
        );
        assert_eq!(shares_for_value(10, 3_000_000).unwrap(), 3);
        assert_eq!(
            share_price_from_nav(1_990_000_000, 995_000_000).unwrap(),
            2_000_000
        );
        assert_eq!(
            value_for_shares(497_500_000, 2_000_000).unwrap(),
            995_000_000
        );
        assert_eq!(management_fee_shares(25_000_000_000).unwrap(), 87_807_325);
        let (half_month, remainder) =
            management_fee_shares_for_elapsed(25_000_000_000, MANAGEMENT_FEE_PERIOD_SECS / 2, 0)
                .unwrap();
        let (second_half, _) = management_fee_shares_for_elapsed(
            25_000_000_000 + half_month,
            MANAGEMENT_FEE_PERIOD_SECS / 2,
            remainder,
        )
        .unwrap();
        assert!(half_month > 0);
        assert!(second_half >= half_month);
    }

    #[test]
    fn realized_profit_fee_and_basis_rounding() {
        assert_eq!(fee_floor(20_000_000, 1_000).unwrap(), 2_000_000);
        assert_eq!(
            cost_basis_for_shares(100_000_001, 50, 100).unwrap(),
            50_000_001
        );
        assert_eq!(
            cost_basis_for_shares(50_000_000, 50, 50).unwrap(),
            50_000_000
        );
        assert_eq!(
            minimum_after_slippage(100_000_000, 200).unwrap(),
            98_000_000
        );
    }
}

File: backend/src/accounting/math.ts
import {
  DEPOSIT_FEE_BPS,
  EARLY_WITHDRAWAL_FEE_BPS,
  MANAGEMENT_FEE_BPS,
  MANAGEMENT_FEE_PERIOD_SECS,
  MATURE_HOLDING_PERIOD_SECS,
  MATURE_WITHDRAWAL_FEE_BPS,
  MAX_BPS,
  MAX_CREATOR_PERFORMANCE_FEE_BPS,
  MAX_MANAGEMENT_FEE_PERIODS,
  SHARE_SCALE,
} from "./constants.js";
import {
  addU128,
  addU64,
  basisPoints,
  i64,
  multiplyU128,
  subU64,
  subtractI64,
  u128,
  u64,
} from "./integers.js";

const MAX_BPS_BIGINT = BigInt(MAX_BPS);

/** Protocol fees round up, matching Rust `fee_ceil`. */
export function feeCeil(amount: bigint, feeBps: number): bigint {
  const checkedAmount = u64(amount, "amount");
  const checkedBps = basisPoints(feeBps, "feeBps");
  const product = multiplyU128(checkedAmount, checkedBps, "fee numerator");
  const numerator = addU128(
    product,
    MAX_BPS_BIGINT - 1n,
    "fee numerator",
  );
  return u64(numerator / MAX_BPS_BIGINT, "fee");
}

/** Performance fees round down, matching Rust `fee_floor`. */
export function feeFloor(amount: bigint, feeBps: number): bigint {
  const checkedAmount = u64(amount, "amount");
  const checkedBps = basisPoints(feeBps, "feeBps");
  const numerator = multiplyU128(
    checkedAmount,
    checkedBps,
    "fee numerator",
  );
  return u64(numerator / MAX_BPS_BIGINT, "fee");
}

/** Deposit share issuance rounds down in the protocol's favour. */
export function sharesForValue(netValue: bigint, sharePrice: bigint): bigint {
  const checkedNet = u64(netValue, "netValue");
  const checkedPrice = u64(sharePrice, "sharePrice");
  if (checkedPrice === 0n) {
    throw new RangeError("sharePrice must be positive");
  }
  const numerator = multiplyU128(
    checkedNet,
    SHARE_SCALE,
    "share numerator",
  );
  return u64(numerator / checkedPrice, "shares");
}

/** Current share price rounds down to six accounting decimals. */
export function sharePriceFromNav(
  navValue: bigint,
  totalShares: bigint,
): bigint {
  const checkedNav = u64(navValue, "navValue");
  const checkedShares = u64(totalShares, "totalShares");
  if (checkedShares === 0n) {
    throw new RangeError("totalShares must be positive");
  }
  const numerator = multiplyU128(
    checkedNav,
    SHARE_SCALE,
    "share-price numerator",
  );
  return u64(numerator / checkedShares, "sharePrice");
}

/** Gross value represented by shares rounds down. */
export function valueForShares(
  shares: bigint,
  sharePrice: bigint,
): bigint {
  const numerator = multiplyU128(
    u64(shares, "shares"),
    u64(sharePrice, "sharePrice"),
    "share-value numerator",
  );
  return u64(numerator / SHARE_SCALE, "shareValue");
}

/** Exact pro-rata final value; rounds down like the on-chain implementation. */
export function proRataValue(
  totalValue: bigint,
  shares: bigint,
  totalShares: bigint,
): bigint {
  const checkedTotalShares = u64(totalShares, "totalShares");
  if (checkedTotalShares === 0n) {
    throw new RangeError("totalShares must be positive");
  }
  const numerator = multiplyU128(
    u64(totalValue, "totalValue"),
    u64(shares, "shares"),
    "pro-rata numerator",
  );
  return u64(numerator / checkedTotalShares, "proRataValue");
}

/**
 * Allocates basis to redeemed shares. A partial redemption rounds basis up in
 * the user's favour; a full redemption consumes the exact remaining basis.
 */
export function costBasisForShares(
  costBasis: bigint,
  sharesRedeemed: bigint,
  sharesOwned: bigint,
): bigint {
  const checkedBasis = u64(costBasis, "costBasis");
  const checkedRedeemed = u64(sharesRedeemed, "sharesRedeemed");
  const checkedOwned = u64(sharesOwned, "sharesOwned");
  if (checkedOwned === 0n || checkedRedeemed > checkedOwned) {
    throw new RangeError("sharesRedeemed must not exceed positive sharesOwned");
  }
  if (checkedRedeemed === checkedOwned) {
    return checkedBasis;
  }
  const product = multiplyU128(
    checkedBasis,
    checkedRedeemed,
    "cost-basis numerator",
  );
  const numerator = addU128(
    product,
    checkedOwned - 1n,
    "cost-basis numerator",
  );
  return u64(numerator / checkedOwned, "withdrawnCostBasis");
}

/** Protocol slippage floor rounds up so tolerance is never exceeded by dust. */
export function minimumAfterSlippage(
  quotedOut: bigint,
  toleranceBps: number,
): bigint {
  const checkedOut = u64(quotedOut, "quotedOut");
  const tolerance = basisPoints(toleranceBps, "toleranceBps");
  const retained = MAX_BPS_BIGINT - tolerance;
  const product = multiplyU128(
    checkedOut,
    retained,
    "slippage numerator",
  );
  const numerator = addU128(
    product,
    MAX_BPS_BIGINT - 1n,
    "slippage numerator",
  );
  return u64(numerator / MAX_BPS_BIGINT, "minimumOutput");
}

/** One full 30-day management-fee period, retained for parity with Rust. */
export function managementFeeShares(totalShares: bigint): bigint {
  const denominator = MAX_BPS_BIGINT - BigInt(MANAGEMENT_FEE_BPS);
  const numerator = multiplyU128(
    u64(totalShares, "totalShares"),
    BigInt(MANAGEMENT_FEE_BPS),
    "management-fee numerator",
  );
  return u64(numerator / denominator, "managementFeeShares");
}

export interface ManagementFeeIntervalResult {
  readonly minted: bigint;
  readonly remainder: bigint;
}

/** Exact time-weighted dilution for one interval. */
export function managementFeeSharesForElapsed(
  totalShares: bigint,
  elapsedSeconds: bigint,
  priorRemainder: bigint,
): ManagementFeeIntervalResult {
  const checkedTotal = u64(totalShares, "totalShares");
  const checkedElapsed = i64(elapsedSeconds, "elapsedSeconds");
  if (checkedElapsed < 0n) {
    throw new RangeError("elapsedSeconds cannot be negative");
  }
  const checkedRemainder = u128(priorRemainder, "priorRemainder");
  const dilutionDenominator = multiplyU128(
    MAX_BPS_BIGINT - BigInt(MANAGEMENT_FEE_BPS),
    MANAGEMENT_FEE_PERIOD_SECS,
    "management-fee denominator",
  );
  if (checkedRemainder >= dilutionDenominator) {
    throw new RangeError("priorRemainder exceeds the dilution denominator");
  }

  let numerator = multiplyU128(
    checkedTotal,
    BigInt(MANAGEMENT_FEE_BPS),
    "management-fee numerator",
  );
  numerator = multiplyU128(
    numerator,
    checkedElapsed,
    "management-fee numerator",
  );
  numerator = addU128(
    numerator,
    checkedRemainder,
    "management-fee numerator",
  );
  return {
    minted: u64(numerator / dilutionDenominator, "managementFeeShares"),
    remainder: numerator % dilutionDenominator,
  };
}

export interface ManagementFeeState {
  readonly totalSharesOutstanding: bigint;
  readonly protocolFeeShares: bigint;
  readonly lastManagementFeeAt: bigint;
  readonly managementFeeAccrualRemainder: bigint;
}

export interface ManagementFeeAccrualResult extends ManagementFeeState {
  readonly mintedShares: bigint;
  readonly elapsedSeconds: bigint;
  readonly completePeriods: bigint;
}

/**
 * Mirrors `accrue_management_fee_internal`, including 30-day chunking,
 * compounding, exact-second accrual and sub-share remainder carry.
 */
export function accrueManagementFee(
  state: ManagementFeeState,
  now: bigint,
): ManagementFeeAccrualResult {
  let totalShares = u64(
    state.totalSharesOutstanding,
    "totalSharesOutstanding",
  );
  let protocolShares = u64(
    state.protocolFeeShares,
    "protocolFeeShares",
  );
  const checkedNow = i64(now, "now");
  const lastAccrual = i64(
    state.lastManagementFeeAt,
    "lastManagementFeeAt",
  );
  let remainder = u128(
    state.managementFeeAccrualRemainder,
    "managementFeeAccrualRemainder",
  );
  const userShares = subU64(
    totalShares,
    protocolShares,
    "user share balance",
  );

  if (totalShares === 0n || userShares === 0n || lastAccrual === 0n) {
    return {
      totalSharesOutstanding: totalShares,
      protocolFeeShares: protocolShares,
      lastManagementFeeAt: checkedNow,
      managementFeeAccrualRemainder: 0n,
      mintedShares: 0n,
      elapsedSeconds: 0n,
      completePeriods: 0n,
    };
  }

  const elapsed = subtractI64(checkedNow, lastAccrual, "elapsed seconds");
  if (elapsed < 0n) {
    throw new RangeError("now cannot precede lastManagementFeeAt");
  }
  if (elapsed === 0n) {
    return {
      totalSharesOutstanding: totalShares,
      protocolFeeShares: protocolShares,
      lastManagementFeeAt: lastAccrual,
      managementFeeAccrualRemainder: remainder,
      mintedShares: 0n,
      elapsedSeconds: 0n,
      completePeriods: 0n,
    };
  }

  const periods = elapsed / MANAGEMENT_FEE_PERIOD_SECS;
  if (periods > MAX_MANAGEMENT_FEE_PERIODS) {
    throw new RangeError("management-fee catch-up exceeds 600 periods");
  }

  let totalMinted = 0n;
  let remaining = elapsed;
  while (remaining > 0n) {
    const interval =
      remaining < MANAGEMENT_FEE_PERIOD_SECS
        ? remaining
        : MANAGEMENT_FEE_PERIOD_SECS;
    const result = managementFeeSharesForElapsed(
      totalShares,
      interval,
      remainder,
    );
    remainder = result.remainder;
    totalShares = addU64(
      totalShares,
      result.minted,
      "totalSharesOutstanding",
    );
    protocolShares = addU64(
      protocolShares,
      result.minted,
      "protocolFeeShares",
    );
    totalMinted = addU64(totalMinted, result.minted, "mintedShares");
    remaining -= interval;
  }

  return {
    totalSharesOutstanding: totalShares,
    protocolFeeShares: protocolShares,
    lastManagementFeeAt: checkedNow,
    managementFeeAccrualRemainder: remainder,
    mintedShares: totalMinted,
    elapsedSeconds: elapsed,
    completePeriods: periods,
  };
}

/** Cost-basis-weighted holding timestamp used by the contract. */
export function weightedAverageTimestamp(
  existingCostBasis: bigint,
  existingTimestamp: bigint,
  addedCostBasis: bigint,
  now: bigint,
): bigint {
  const existingBasis = u64(existingCostBasis, "existingCostBasis");
  const addedBasis = u64(addedCostBasis, "addedCostBasis");
  const checkedExistingTimestamp = i64(
    existingTimestamp,
    "existingTimestamp",
  );
  const checkedNow = i64(now, "now");
  if (addedBasis === 0n || checkedNow <= 0n) {
    throw new RangeError("addedCostBasis and now must be positive");
  }
  if (existingBasis === 0n) {
    return checkedNow;
  }
  if (
    checkedExistingTimestamp <= 0n ||
    checkedExistingTimestamp > checkedNow
  ) {
    throw new RangeError("existingTimestamp is invalid");
  }

  const totalBasis = addU64(existingBasis, addedBasis, "totalCostBasis");
  const existingWeighted = multiplyU128(
    checkedExistingTimestamp,
    existingBasis,
    "weighted timestamp numerator",
  );
  const addedWeighted = multiplyU128(
    checkedNow,
    addedBasis,
    "weighted timestamp numerator",
  );
  const numerator = addU128(
    existingWeighted,
    addedWeighted,
    "weighted timestamp numerator",
  );
  return i64(numerator / totalBasis, "weightedDepositTimestamp");
}

export interface WithdrawalBreakdown {
  readonly costBasis: bigint;
  readonly realizedProfit: bigint;
  readonly earlyExitValue: bigint;
  readonly matureExitValue: bigint;
}

export function weightedAverageWithdrawalBreakdown(
  costBasisValue: bigint,
  sharesOwned: bigint,
  weightedDepositTimestamp: bigint,
  sharesToConsume: bigint,
  grossValue: bigint,
  now: bigint,
): WithdrawalBreakdown {
  const timestamp = i64(
    weightedDepositTimestamp,
    "weightedDepositTimestamp",
  );
  const checkedNow = i64(now, "now");
  if (timestamp <= 0n || timestamp > checkedNow) {
    throw new RangeError("weightedDepositTimestamp is invalid");
  }
  const costBasis = costBasisForShares(
    costBasisValue,
    sharesToConsume,
    sharesOwned,
  );
  const checkedGross = u64(grossValue, "grossValue");
  const age = subtractI64(checkedNow, timestamp, "holding age");
  const mature = age >= MATURE_HOLDING_PERIOD_SECS;
  return {
    costBasis,
    realizedProfit:
      checkedGross > costBasis ? checkedGross - costBasis : 0n,
    earlyExitValue: mature ? 0n : checkedGross,
    matureExitValue: mature ? checkedGross : 0n,
  };
}

export interface DepositSettlementMath {
  readonly protocolFee: bigint;
  readonly quotedNetValue: bigint;
  readonly minimumNetValue: bigint;
  readonly sharesCredited: bigint;
}

export function calculateDepositSettlement(
  grossAmount: bigint,
  actualNetValue: bigint,
  sharePrice: bigint,
  maxSlippageBps: number,
): DepositSettlementMath {
  const protocolFee = feeCeil(grossAmount, DEPOSIT_FEE_BPS);
  const quotedNetValue = subU64(
    grossAmount,
    protocolFee,
    "quotedNetValue",
  );
  const checkedActual = u64(actualNetValue, "actualNetValue");
  const minimumNetValue = minimumAfterSlippage(
    quotedNetValue,
    maxSlippageBps,
  );
  if (checkedActual > quotedNetValue || checkedActual < minimumNetValue) {
    throw new RangeError("actualNetValue is outside the allowed range");
  }
  const sharesCredited = sharesForValue(checkedActual, sharePrice);
  if (sharesCredited === 0n) {
    throw new RangeError("deposit would credit zero shares");
  }
  return {
    protocolFee,
    quotedNetValue,
    minimumNetValue,
    sharesCredited,
  };
}

export interface WithdrawalSettlementMath extends WithdrawalBreakdown {
  readonly creatorFee: bigint;
  readonly protocolFee: bigint;
  readonly userValueOut: bigint;
}

export function calculateWithdrawalSettlement(
  costBasisValue: bigint,
  sharesOwned: bigint,
  weightedDepositTimestamp: bigint,
  sharesToConsume: bigint,
  grossValue: bigint,
  now: bigint,
  performanceFeeBps: number,
): WithdrawalSettlementMath {
  if (
    !Number.isInteger(performanceFeeBps) ||
    performanceFeeBps < 0 ||
    performanceFeeBps > MAX_CREATOR_PERFORMANCE_FEE_BPS
  ) {
    throw new RangeError(
      `performanceFeeBps must be in [0, ${MAX_CREATOR_PERFORMANCE_FEE_BPS}]`,
    );
  }
  const breakdown = weightedAverageWithdrawalBreakdown(
    costBasisValue,
    sharesOwned,
    weightedDepositTimestamp,
    sharesToConsume,
    grossValue,
    now,
  );
  const creatorFee = feeFloor(
    breakdown.realizedProfit,
    performanceFeeBps,
  );
  const earlyFee = feeCeil(
    breakdown.earlyExitValue,
    EARLY_WITHDRAWAL_FEE_BPS,
  );
  const matureFee = feeCeil(
    breakdown.matureExitValue,
    MATURE_WITHDRAWAL_FEE_BPS,
  );
  const protocolFee = addU64(earlyFee, matureFee, "protocolFee");
  const afterProtocol = subU64(grossValue, protocolFee, "userValueOut");
  const userValueOut = subU64(afterProtocol, creatorFee, "userValueOut");
  return { ...breakdown, creatorFee, protocolFee, userValueOut };
}
